linux/src/threadpool_linux.c

```diff
@@ -501,29 +531,156 @@ int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCT
+                LogError("Failure calling timer_create. Error: %s", MU_P_OR_NULL(err_msg));
+            }
+            else
+            {
+                struct itimerspec its = {0};
+                its.it_value.tv_sec = start_delay_ms / 1000;
+                its.it_value.tv_nsec = start_delay_ms * MILLISEC_TO_NANOSEC % 1000000000;
+                its.it_interval.tv_sec = 0;
+                its.it_interval.tv_nsec = timer_period_ms * MILLISEC_TO_NANOSEC;
+
+                if (timer_settime(time_id, 0, &its, NULL) == -1)
+                {
+                    char err_msg[128] = {0};
+                    (void)strerror_r(errno, err_msg, 128);
+                    LogError("Failure calling timer_settime. Error: %s", MU_P_OR_NULL(err_msg));
+                }
+                else
+                {
+                    (void)interlocked_decrement(&threadpool->pending_call_count);
+                    wake_by_address_all(&threadpool->pending_call_count);
+                    timer_instance->time_id = time_id;
+                    *timer_handle = timer_instance;
+                    result = 0;
+                    goto all_ok;
+                }
+            }
+            free(timer_instance);
+        }
+        result = MU_FAILURE;
+        (void)interlocked_decrement(&threadpool->pending_call_count);
+        wake_by_address_all(&threadpool->pending_call_count);
+    }
+all_ok:
+    return result;
 }
 
 int threadpool_timer_restart(TIMER_INSTANCE_HANDLE timer, uint32_t start_delay_ms, uint32_t timer_period_ms)
 {
-    (void)timer;
-    (void)start_delay_ms;
-    (void)timer_period_ms;
-    return MU_FAILURE;
+    int result;
+    if (
+        timer == NULL
+        )
+    {
+        LogError("Invalid args: TIMER_INSTANCE_HANDLE timer = %p, uint32_t start_delay_ms = %" PRIu32 ", uint32_t timer_period_ms = %" PRIu32 "",
+            timer, start_delay_ms, timer_period_ms);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        struct itimerspec its = {0};
+        its.it_value.tv_sec = start_delay_ms / 1000;
+        its.it_value.tv_nsec = start_delay_ms * MILLISEC_TO_NANOSEC % 1000000000;
+        its.it_interval.tv_sec = timer_period_ms / 1000;
+        its.it_interval.tv_nsec = timer_period_ms * MILLISEC_TO_NANOSEC % 1000000000;
+
+        if (timer_settime(timer->time_id, 0, &its, NULL) != 0)
+        {
+            char err_msg[128] = {0};
+            (void)strerror_r(errno, err_msg, 128);
+            LogError("Failure calling timer_settime. Error: %s", MU_P_OR_NULL(err_msg));
+            result = MU_FAILURE;
+        }
+        else
+        {
+            result = 0;
+        }
+    }
+    return result;
 }
 
 void threadpool_timer_cancel(TIMER_INSTANCE_HANDLE timer)
 {
-    (void)timer;
+    if (timer == NULL)
+    {
+        LogError("Invalid args: TIMER_INSTANCE_HANDLE timer = %p", timer);
+    }
+    else
+    {
+        struct itimerspec its;
+        memset(&its, 0, sizeof(its));
```

Comment:
can we use a global here initialized to "all 0" as opposed to build it every time?